### PR TITLE
Approach task

### PR DESCRIPTION
**Pull request by liatsade**
Training on entities - how to approach a new mission.

https://www.gitbook.com/book/amitaibu/the-gizra-way/edit#/edit/approach_mission/approach_mission.md

### CODE REVIEW

**Comment by liatsade** [comment]
@RachelBaram post ready to review. 
Please notice to:
What is the best way to represent the Entity in word? I put it in quotes but maybe there is more understandable way...

wip__approach_mission_by_liatsade_ _pull_request__41_ _gizra_gizra-way-book

**Comment by liatsade** [comment]
@amitaibu 
wip__approach_mission_by_liatsade_ _pull_request__41_ _gizra_gizra-way-book

Maybe we can allowed one article belongs to many topics, and explain that in this situation we can still choose the article to refer to topic, because any case it won't be a million topics for one article.

draw for one article ->one topic
wip__approach_mission_by_liatsade_ _pull_request__41_ _gizra_gizra-way-book

draw for one article -> many topics
approach_mission_md__approach_mission__-_gitbook_editor

File: approach_mission.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+In that case “Membership” refer to “Topic” (because we don’t want “Topic” refer to million “Membership”) and ““Membership” also refer to “User” (because we don’t want “User” refer to million “Membership”). It looks like this:
+
+![](88.jpg)
+
+
+
+Let us continue with our story - 
+
+**Registration expired after a year. The user will get a reminder emails - 3 month, 1 month and 1 day before expired, so he can renew his registration.**
```
Comment (amitaibu) [inline approach_mission.md]: I usually prefer to highlight just a word or two. Once it's a whole sentence it's hard to follow on what's really important.

File: approach_mission.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+![](4.jpg)
+
+
+And who refer to whom? (what is the arrow direction?)
+Let’s ask the “million question” - 
+Is one topic can has more than one article? Yes.
+Is one article can written on more than one topic? Theoretically yes, but for the practice we decide that every article belongs only to one topic, so the answer to this question is No.
+
+
+//pic5a
```
Comment (amitaibu) [inline approach_mission.md]: leftovers?

File: approach_mission.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+![](11.jpg)
+
+
+Finally, let's describe the exact query:
+
+**Give me all “membership” that their “status” is *active*, and their “time stamp” is *today's date minus 9 month*, and we didn't sent email yet to the user who created this membership.**
+
+
+Summary:
```
Comment (amitaibu) [inline approach_mission.md]: In the summary I would put all the "rules" you went over in the article

File: approach_mission.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+Ok, so now we have all the information we need for sending the first reminder email (the one that comes after 9 month), so in order to know to which user we need to send that email, we need to retrieve from the database the right user. 
+Let’s describe in words the query for getting the right information from database:
+
+**Give me all “membership” that their “status” is *active* and their “time stamp” is *today's date minus 9 month***
+
+Ok, we almost there, but there is still one more thing - we need to check that we don't send the email more then one time to the same user. Assuming that we have a lots of "membership" in our database, so we need to limit the number of "membership" we get every time we run the query (if the system will bring us all the fit "membership" at ones, it can be run out of memory). So we will tell the system to brings us only 100 "membership" at a time, and we will run the query every 5 min. 
+How we make sure that the system won't bring us the same "membership" we already sent email for them 5 min ago?
+Of course we need to save an Email log, and check it every time we want to send reminder email.
+
+so let's add new entity call "email log"
```
Comment (amitaibu) [inline approach_mission.md]: if you have machine names such as email log I think it's better to highlight as code (i.e. `email log`)

**Comment by liatsade** [comment]
@RachelBaram I did some fixes in the text, and also add summary. 
It finished from my side and waiting for your loving hands, make it write in proper English :-)
Thanks!

**Comment by RachelBaram** [comment]
@liatsade  So far I have edited up to the section about membership. I did some content changes to try to stay consistent with the two question approach. Please take a look and let me know if this works for you. If it does, I suggest continuing this way of representing the logic (without shortcuts) to help drill in the concept (repetition of the 2 questions at all of the relevant steps). Once I get your feedback about this, I will continue editing. ;)

**Comment by liatsade** [comment]
@RachelBaram 
The repetition of the 2 questions is very good! I add it also to the last section (about email log).
I liked that you replaced the "he" with "she" :-)
Please continue. Thanks.

File: approach_mission.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+How do we make sure that the system won't bring us the same `membership` we already sent email for 5 min ago?
+
+We need to save the information for which `membership` we already sent emails, and check it every time we want to send reminder email.
+But where are we going to save this information?
+Is it going to be at the `user` entity? Well, remember that `user` can have more than one `membership`, so it will be very complex to put all this data at the `user` entity. Additionally, the email sends based on a membership's timestamp, so it makes more sense that the `membership` will hold the information about the emails who was send. Well, this won't be good enough ether, because we need to send three emails (3 month, 1 month and 1 day before the membership expire), and it's to much data to put at at the `membership` entity.
+
+In that case it will be better to create new entity. Let's call it `email log`.
```
Comment (liatsade) [inline approach_mission.md]: @amitaibu please review this new part about email log, I want to make sure the explanation is correct before @RachelBaram do English review.

**Comment by liatsade** [comment]
@RachelBaram I did some changes in the content, please review.
- took out the "plot" idea - please make sure the explanation still understandable without it.
- expanded the explantion about email log - @amitaibu is it ok now?
- expanded the summary.

File: approach_mission.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+Take a look at the following simple story: 
+
+
+> An author writes articles.
+
+
+Let's try to identify the entities and relationships in this story. The best way to do it - is to draw it! Use circles for entities and lines for relationships.
+
+![](111.jpg)
```
Comment (amitaibu) [inline approach_mission.md]: can you change the file names to make more sense (`111`, and `888`). You can also give them names if you prefer

File: approach_mission.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
+# Approaching a Task - determining a website's technical structure.
```
Comment (amitaibu) [inline approach_mission.md]: lets change the file name as well to `approach_data_structure_task`

How about the title `How to approach a new task, and its underlying data structure`

**Comment by amitaibu** [comment]
when ready for review, please remove the `WIP` prefix

File: approach_mission.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+Up until now nothing was too complicated. Even if you are not a web developer, you are probably comfortable with the above logic. You might be tempted to think that the below requirements require some deep understanding of computer science. They don't!
+
+The simplicity and logic presnted can assist us even with the next scenarios:
+
+> We are going to build a Premium Website that contains articles. People can sign-up and register to one or more topics of interest. Registering to a topic means that they can read articles that belong to it.
+
+In the above example we cheated a bit since we have called the `User` entity an `Author`. But it's a real user. It's very rare of a site not to have a user, since, well you know - we, the people, are still important.
+
+
+Question 1: Does a `user` have a relationship with a `topic`?
```
Comment (amitaibu) [inline approach_mission.md]: I've edited up until here

**Comment by amitaibu** [comment]
(I've deleted file `88.jpg` as it was a bit too messy)

File: approach_mission.md (hunk 6 above)
Comment (liatsade) [reply approach_mission.md]: I can't find a way to change file name *.jpg in github. I think it needs to be change locally and re-upload it.
I want to upload new sketch anyway, so I'll do it then.

File: approach_data_structure_task.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+![](2.jpg)
+
+And, if our author is particularly diligent, she can even write a million articles (theoretically of course), so this brings us to the next important question:
+
+**Question 2. The Million Question** Can and should a single `author` entity refer to a million `article` entities?
+
+Answer: No, the answer is definitely not. It would be a bad idea. 
+
+To understand this, we need to talk about the **meaning of reference**. When object A refers to object B it means object A “knows” about object B. This "knowing" becomes part of the information that object A holds.  So for example when we call object A from the database, it might also retrieve the information of all the "B" objects that it refers to. So, if theoretically there can be a million "B" objects, the system would retrieve them all. This is a very heavy task for the system that requires a lot of memory resources. We want to avoid this.
```
Comment (RachelBaram) [inline approach_data_structure_task.md:35]: @liatsade I think a more real example (instead of A and B) will get the point across better.

File: approach_data_structure_task.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+ 
+What information do we need in order to send the reminder emails?
+
+We need to know, of course, when the membership began, so we can calculate when 9 months has passed to send the first reminder email. In technical words, the beginning date or any other date that represent that something has occured in the system is called a `timestamp`.
+
+But this information is not enough. Think about a situation where the user decides not to renew his registration. What happened to his membership? Does it go away, get deleted from the system? Well, usually we don’t delete content from the system, but we can mark it as inactive.
+
+So in order to send the user a reminder, we need to know two things:  the membership created date (`timestamp`) and if it is active or not (`status`).
+
+![](9.jpg)
```
Comment (RachelBaram) [inline approach_data_structure_task.md]: @liatsade We have decided that circles are for entities. I think putting timestamp and status in a circle is confusing. If that is how things are done, then we should explain it.

Comment (amitaibu) [reply approach_data_structure_task.md]: Indeed, shouldn't as circles - because they are a `property` of the `entity`. Can be in a follow up, for when we replace the images to something like we have in http://www.gizra.com/content/og8-development-mindset/

Comment (liatsade) [reply approach_data_structure_task.md]: I added it to issue #44

File: approach_data_structure_task.md (hunk 9 above)
Comment (liatsade) [reply approach_data_structure_task.md:35]: The example is in the content itself (author and article), I meant here to explain the principle in general.
I added sentence after this section that link between the principle and our case - is it clarify that?

**Comment by RachelBaram** [comment]
@liatsade  I made some content changes in the second half of the document (including the summary). Let me know if this works. I tried to keep the logic steps consistent and clear. Once you and Amitai finish up your changes, I will do a final proofreading.

File: approach_data_structure_task.md (hunk 9 above)
Comment (RachelBaram) [reply approach_data_structure_task.md:35]: I don't think that will help. What helped me to understand is the list of properties (that is why the circles around none entities bothered me). Literally saying something like, "The article will hold the author's name. The author will not hold the names of all the articles."

File: approach_data_structure_task.md (hunk 10 above)
Comment (RachelBaram) [reply approach_data_structure_task.md]: @amitaibu I like the less exact circles. ;) There is an ease to it.  People can totally see themselves sketching like this. I keep on imagining people trying to get their pens around that bottle cap that keeps on moving on the page.

Comment (amitaibu) [reply approach_data_structure_task.md]: Look carefully at the images in the post - I can assure you the page has moved 😉

File: approach_data_structure_task.md (hunk 9 above)
Comment (liatsade) [reply approach_data_structure_task.md:35]: Good! Let's keep the "A" and "B" explanation, and add your sentence after it, so it will do the link to the case we talk about.

**Comment by liatsade** [comment]
Looks good. I liked you add the "entity A/B" to summary, it's more clear like that :-)

**Comment by amitaibu** [comment]
Images are located in the wrong place, I'll move them.

10_jpg_ __library_webserver_documents_the-gizra-way

**Comment by amitaibu** [comment]
GitBook returns 500 error, lets see if the merge will work.

**Comment by amitaibu** [comment]
Merged, thanks.

Lets tackle next the images, so this one will be complete

**Comment by liatsade** [comment]
Thanks!
Just me to know for next time - this time I uploaded the images by gitbook UI:
approach_data_structure_task_md__master__-_gitbook_editor

There is no place to determine where the image will be locate. 
@amitaibu what is the best way to upload images?

**Comment by liatsade** [comment]
I found two little typo mistake, I fixed them directly on Master.

**Comment by amitaibu** [comment]
> @amitaibu what is the best way to upload images?

If you can't move it via the UI (and I think you can), you can do it locally and use git commit/ push
